lib/policies/deserializationPolicy.ts

```diff
@@ -37,109 +38,133 @@ export class DeserializationPolicy extends BaseRequestPolicy {
   }
 }
 
-function deserializeResponseBody(response: HttpOperationResponse): Promise<HttpOperationResponse> {
-  return parse(response).then(response => {
-    const operationSpec: OperationSpec | undefined = response.request.operationSpec;
-    if (operationSpec && operationSpec.responses) {
-      const statusCode: number = response.status;
-
-      const expectedStatusCodes: string[] = Object.keys(operationSpec.responses);
-
-      const hasNoExpectedStatusCodes: boolean = (expectedStatusCodes.length === 0 || (expectedStatusCodes.length === 1 && expectedStatusCodes[0] === "default"));
-
-      const responseSpec: OperationResponse = operationSpec.responses[statusCode];
-
-      const isExpectedStatusCode: boolean = hasNoExpectedStatusCodes ? (200 <= statusCode && statusCode < 300) : !!responseSpec;
-      if (!isExpectedStatusCode) {
-        const defaultResponseSpec: OperationResponse = operationSpec.responses.default;
-        if (defaultResponseSpec) {
-          const initialErrorMessage: string = isStreamOperation(operationSpec.responses)
-            ? `Unexpected status code: ${statusCode}`
-            : response.bodyAsText as string;
-
-          const error = new RestError(initialErrorMessage);
-          error.statusCode = statusCode;
-          error.request = utils.stripRequest(response.request);
-          error.response = utils.stripResponse(response);
-
-          let parsedErrorResponse: { [key: string]: any } = response.parsedBody;
-          try {
-            if (parsedErrorResponse) {
-              const defaultResponseBodyMapper: Mapper | undefined = defaultResponseSpec.bodyMapper;
-              if (defaultResponseBodyMapper && defaultResponseBodyMapper.serializedName === "CloudError") {
-                if (parsedErrorResponse.error) {
-                  parsedErrorResponse = parsedErrorResponse.error;
-                }
-                if (parsedErrorResponse.code) {
-                  error.code = parsedErrorResponse.code;
-                }
-                if (parsedErrorResponse.message) {
-                  error.message = parsedErrorResponse.message;
-                }
-              } else {
-                let internalError: any = parsedErrorResponse;
-                if (parsedErrorResponse.error) {
-                  internalError = parsedErrorResponse.error;
-                }
+async function getOperationResponse(parsedResponse: HttpOperationResponse): Promise<undefined | OperationResponse> {
```

Comment:
where's the async part of this?

Reply:
Line 50. Is this one of those cases where it's better to not use async/await due to code size constraints?

Reply:
Avoiding async/await in the non-test code is good to minimize size.

More importantly though I'm struggling to see why `operationResponseGetter` sometimes returns a Promise and other times doesn't. Is it because it sometimes needs to choose the response spec based the result of some async operation?

Reply:
I just took another look and realized that we must've changed getting the response body to a synchronous operation/variable a while ago. I was still under the impression that was an async operation. I'll remove the Promise return option.

Reply:
OK, it seems like that should hopefully simplify the types of several methods and variables, and make the await unnecessary.